model/src/main/scala/hmda/model/institution/Institution.scala

```diff
@@ -0,0 +1,53 @@
+package hmda.model.institution
+
+/**
+ * A financial institution, geared towards requirements for filing HMDA data.
+ */
+case class Institution(
+    id: Int,
+    name: String,
+    externalIds: Set[ExternalId],
+    agency: Agency,
+    institutionType: InstitutionType
+) {
+
+  val extIdsByType: Map[ExternalIdType, ExternalId] = externalIds.map(extId => (extId.idType, extId)).toMap
+
+  /**
+   * Derives the respondentId for a given Institution based on [[hmda.model.institution.Agency]] and [[hmda.model.institution.InstitutionType]],
+   * the rules for which can be found in section "1.4 - Respondent Identification Numbers for 2017 HMDA Filers" of the
+   * <a href="http://www.consumerfinance.gov/data-research/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf">2017 HMDA File Specifications</a>
+   */
+  def respondentId: Either[InvalidRespondentId, ExternalId] = {
+
+    institutionType.depositoryType match {
+      case None => Left(NoDepositoryTypeForInstitutionType(id, institutionType))
+      case Some(dt) =>
+
+        agency.externalIdTypes.get(dt) match {
+          case None => Left(UnsupportedDepositoryTypeByAgency(id, agency, dt))
+          case Some(extIdType) =>
+
+            extIdsByType.get(extIdType) match {
+              case None => Left(RequiredExternalIdNotPresent(id, extIdType))
+              case Some(extId) => Right(ExternalId(extId.id, extIdType))
+            }
+        }
+    }
```

Comment:
I agree, pattern matching on `Option` is usually not want you want to do. A more idiomatic version of this function: 

```scala
  def respondentId: Either[InvalidRespondentId, ExternalId] = {
    institutionType.depositoryType.map { dt =>
      agency.externalIdTypes.get(dt)
        .map { extIdType =>
          extIdsByType.get(extIdType)
            .map(extId => Right(ExternalId(extId.id, extIdType)))
            .getOrElse(Left(RequiredExternalIdNotPresent(id, extIdType)))
        }
        .getOrElse(Left(UnsupportedDepositoryTypeByAgency(id, agency, dt)))
    }.getOrElse(Left(NoDepositoryTypeForInstitutionType(id, institutionType)))
  }
```

As @lgdean has pointed out, if you don't need to get error values this translates nicely to a `for-comprehension`. I'm also a bit suspicious of the `Either` usage here. Is it really necessary? I wonder if an `Option[RespondentId]` might work better